simplex/storage.go

```diff
@@ -0,0 +1,239 @@
+	lastAccepted, err := config.VM.LastAccepted(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	lastAcceptedBlock, err := config.VM.GetBlock(ctx, lastAccepted)
+	if err != nil {
+		return nil, err
+	}
+
+	genesisBlock, err := getGenesisBlock(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Storage{
+		db:           config.DB,
+		genesisBlock: genesisBlock,
+		vm:           config.VM,
+		deserializer: qcDeserializer,
+		blockTracker: blockTracker,
+		log:          config.Log,
+	}
+	s.height.Store(lastAcceptedBlock.Height() + 1)
+
+	return s, nil
+}
+
+func (s *Storage) Height() uint64 {
+	return s.height.Load()
+}
+
+// Retrieve returns the block and finalization at [seq].
+// If [seq] is not found, returns false.
+func (s *Storage) Retrieve(seq uint64) (simplex.VerifiedBlock, simplex.Finalization, bool) {
+	if seq == 0 {
+		return s.genesisBlock, simplex.Finalization{}, true
+	}
+
+	block, err := getBlockAtHeight(context.TODO(), s.vm, seq)
+	if err != nil {
+		if err == database.ErrNotFound {
+			return nil, simplex.Finalization{}, false
+		}
+		return nil, simplex.Finalization{}, false
+	}
+
+	finalization, found, err := s.retrieveFinalization(seq)
+	if !found || err != nil {
+		return nil, simplex.Finalization{}, false
+	}
+
+	vb := &Block{vmBlock: block, metadata: finalization.Finalization.ProtocolMetadata}
+
+	return vb, finalization, true
+}
+
+// Index indexes the finalization in the storage.
+// It stores the finalization bytes at the current height and increments the height.
+func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finalization simplex.Finalization) error {
+	bh := block.BlockHeader()
+	if bh.Seq == 0 {
+		s.log.Warn("attempted to index genesis block, which should not be indexed")
+		return errGenesisIndexed
+	}
+
+	currentHeight := s.height.Load()
+	if currentHeight != bh.Seq {
+		return fmt.Errorf("%w: expected %d, got %d", errUnexpectedSeq, currentHeight, bh.Seq)
+	}
+
+	if !bytes.Equal(bh.Digest[:], finalization.Finalization.Digest[:]) {
+		return fmt.Errorf("%w: expected %d, got %d", errMismatchedDigest, bh.Digest, finalization.Finalization.Digest)
+	}
+
+	if finalization.QC == nil {
+		return errInvalidQC
+	}
+
+	seqBuff := make([]byte, 8)
+	binary.BigEndian.PutUint64(seqBuff, currentHeight)
+
+	finalizationBytes := finalizationToBytes(finalization)
+	if err := s.db.Put(seqBuff, finalizationBytes); err != nil {
+		return fmt.Errorf("failed to store finalization: %w", err)
+	}
+
+	err := s.blockTracker.indexBlock(ctx, bh.Digest)
+	if err != nil {
+		return fmt.Errorf("failed to index block: %w", err)
+	}
+
+	s.height.Add(1) // only increment height after successful indexing
+	return nil
+}
+
+// getGenesisBlock returns the genesis block wrapped as a Block instance.
+func getGenesisBlock(ctx context.Context, config *Config) (*Block, error) {
+	genesis := &Block{
```

Comment:
This comment is unrelated to this line, but, I noticed that we only populate the block tracker in the test. Yet we use this field in production code.... Isn't its initialization missing somewhere?